### PR TITLE
Add expected withdrawals API

### LINKED ISSUE #12261: Fetch withdrawals from a parent state
We should implement https://github.com/ethereum/beacon-APIs/pull/304

### PR DESCRIPTION
**Pull request by tanhengyeow**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR adds a withdrawal endpoint to fetch withdrawals from a parent state. The corresponding PR for the beacon API specs can be found [here](https://github.com/ethereum/beacon-APIs/pull/304).

**Which issues(s) does this PR fix?**

Fixes #12261

**Other notes for review**
- This PR implements the API without using gRPC, similar to the rewards server.
- Upper bound for proposal slot is 4 epochs after the epoch of the current state, which corresponds to `MAX_SEED_LOOKAHEAD`. Teku does it this way too but it is open for discussion :)

### CODE REVIEW

File: beacon-chain/rpc/eth/builder/handlers.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	if stateId == "" {
+		network.WriteError(w, &network.DefaultErrorJson{
+			Message: "state_id is required in URL params",
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+	st, err := s.Stater.State(r.Context(), []byte(stateId))
+	if err != nil {
+		network.WriteError(w, handleWrapError(err, "could not retrieve state_id", http.StatusNotFound))
```
Comment (rkapka) [inline beacon-chain/rpc/eth/builder/handlers.go:31]:
```suggestion
		network.WriteError(w, handleWrapError(err, "could not retrieve state", http.StatusNotFound))
```

File: beacon-chain/rpc/eth/builder/handlers.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	if proposalSlot <= st.Slot() {
+		network.WriteError(w, &network.DefaultErrorJson{
+			Message: fmt.Sprintf("proposal slot must be bigger than state slot. proposal slot: %d, state slot: %d", proposalSlot, st.Slot()),
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+	lookAheadLimit := uint64(params.BeaconConfig().SlotsPerEpoch.Mul(uint64(params.BeaconConfig().MaxSeedLookahead)))
+	// SafeSub not needed as we have checked that proposal slot cannot be lesser than Capella start epoch
+	if proposalSlot.Sub(lookAheadLimit) >= st.Slot() {
```
Comment (rkapka) [inline beacon-chain/rpc/eth/builder/handlers.go:68]: According to the error message this should be a `>`

**Review by potuz — COMMENTED** [review]
Just a couple of comments, thanks for the PR, it's well scoped, well written, easy to follow and a complete work. Looking forward to reading more like this!

File: beacon-chain/rpc/eth/builder/handlers_test.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	require.NoError(t, err)
+	capellaSlot, err := slots.EpochStart(params.BeaconConfig().CapellaForkEpoch)
+	require.NoError(t, err)
+	currentSlot := capellaSlot + primitives.Slot(slotsAhead)
+	require.NoError(t, st.SetSlot(currentSlot))
+	mockChainService := &mock.ChainService{Optimistic: true}
+
+	t.Run("get correct expected withdrawals", func(t *testing.T) {
+		saved := params.BeaconConfig().MaxValidatorsPerWithdrawalsSweep
+		params.BeaconConfig().MaxValidatorsPerWithdrawalsSweep = 16
```
Comment (rkapka) [inline beacon-chain/rpc/eth/builder/handlers_test.go:118]: We usually use something like this in tests:
```
params.SetupTestConfigCleanup(t)
cfg := params.MainnetConfig().Copy()
cfg.ConfigName = "test"
params.OverrideBeaconConfig(cfg)
```

File: beacon-chain/rpc/eth/builder/handlers.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	root, err := helpers.BlockRootAtSlot(st, st.Slot()-1)
+	if err != nil {
+		network.WriteError(w, handleWrapError(err, "could not get block root", http.StatusInternalServerError))
+		return
+	}
+	var blockRoot [32]byte
+	copy(blockRoot[:], root)
+	isFinalized := s.FinalizationFetcher.IsFinalized(r.Context(), blockRoot)
+	// Advance state forward to proposal slot
+	st, err = transition.ProcessSlots(r.Context(), st, proposalSlot)
```
[the remark this answers is not in the document]
Comment (tanhengyeow) [reply beacon-chain/rpc/eth/builder/handlers.go:90]: The endpoint is targeted at external block builders. Quoting from this [issue](https://github.com/ethereum/beacon-APIs/issues/301):
`With this API, withdrawals in the payload of externally built blocks can be compared with the expected withdrawals to ensure that the two match.`

File: beacon-chain/rpc/eth/builder/handlers.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/network"
+	enginev1 "github.com/prysmaticlabs/prysm/v4/proto/engine/v1"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	"net/http"
+	"strconv"
+)
+
+// ExpectedWithdrawals Get the withdrawals computed from the specified state, that will be included in the block that gets built on the specified state.
```
Comment (rkapka) [inline beacon-chain/rpc/eth/builder/handlers.go:20]:
```suggestion
// ExpectedWithdrawals get the withdrawals computed from the specified state, that will be included in the block that gets built on the specified state.
```

**Review by terencechain — APPROVED** [review]
lgtm, really nice PR. I'm a fan. Thank you!

**Review by rkapka — APPROVED** [review]

**Pull request merged**